---
title: "TSS profiles protein-coding transcripts"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("config")
```

`r print(date())`

# Params

```{r label="params"}
```


```{r label="blacklist"}
options(scipen=999)

# https://zenbu-wiki.gsc.riken.jp/zenbu/wiki/index.php/Uploading_UCSC_repetitive_elements_track
# downloading UCSC rmsk data as BED
# BED formatted UCSC track content can be obtained from UCSC table broswer.
# The rmsk RepeatMasker (rmsk) track can be exported as BED file by selecting
#
# the assembly "Feb.2009 GRCh37/hg19"
# the group "repeats and variations"
# the track "RepeatMasker"
# and finally the table "rmsk"
# As we desire the complete repetitive elements genome-wide to be loaded into ZENBU, therefore we select
#
# region: "genome"
# ZENBU enable gzip compressed bed files to be loaded directly, so we will further select :
#
# output format: "BED - browser extensibke format"
# output file: we will name the file "UCSC_rmsk.hg19.bed.gz"
# file type returned: "gzip compressed"

if (!exists("rmsk_raw")) {
  rmsk_raw = read.table("~/projects/small_structs/data/rmsk.mm10.bed", skip=1, stringsAsFactors=FALSE)
}
rmsk = rmsk_raw

# blacklist = rmsk
blacklist = rmsk[rmsk[,4] %in% c("GSAT_MM","SYNREP_MM"),]#, "(CTGTG)n"),]#
blacklist = rbind(blacklist, list("chr19", 61266781, 61266782, "chr19:61266781-61266782", 1404, "+"))
blacklist = rbind(blacklist, list("chr11", 3126500, 3200500, "Sfi1", 300, "-"))
# sat2_feat_saf = rbind(sat2_feat_saf, list("Sfi1", "chr11", 3126500, 3200500, "-"))
print(head(blacklist))
dim(blacklist)
blacklist_filename = "blacklist_file.bed"
write.table(blacklist, file=blacklist_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)    
```



# Features (TSS)


```{r label="tss_feature"}
prefix = "tss"
feat = readRDS("~/projects/genes/bed_grcm38_transcripts_epimeddb_entrez.rds")
feat = feat[!is.na(feat$tx_end),]
feat = feat[nchar(feat$chrom_text) <=5,]
dim(feat)
table(feat$type, useNA="ifany")
table(feat$chrom_text, useNA="ifany")

# tss
table(feat$strand, useNA="ifany")
feat$tss = NA
feat[feat$strand == "+",]$tss = feat[feat$strand == "+",2]
feat[feat$strand == "-",]$tss = feat[feat$strand == "-",3]
feat$tss_key = paste0(feat$chrom_text, "_", feat$tss)#, "_", feat$strand)
sum(duplicated(feat$tss_key))
feat = feat[!duplicated(feat$tss_key),]
dim(feat)

unique(feat[,1])

feat[feat[,1]=="chrMT",1] = "chrM"
unique(feat[,1])


feat = feat[feat$type%in%"protein-coding",]
regions_filenames = regions_filename = paste0(prefix, ".bed")
write.table(feat, file=regions_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)

head(feat)
dim(feat)

feat_saf = feat[,c(5,1:3, 6)]
colnames(feat_saf) = c("GeneID", "Chr", "Start", "End", "Strand")
feat_saf$Start = feat$tss - 5000
feat_saf[feat_saf$Start<1,]$Start = 1
feat_saf$End = feat$tss + 5000
head(feat_saf)
saf_filename = paste0(prefix, ".saf")
write.table(feat_saf, file=saf_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
```

```{r label="featureCounts", eval=FALSE}
mmq=30
cmd = "/summer/epistorage/miniconda3/envs/subread_env/bin/featureCounts"
args = paste0(
  "-a ", prefix, ".saf -F SAF -s 0 -Q ", mmq, " -T 8 -o ", prefix, "_count_", mmq, ".txt ", 
  paste0("/home/fchuffar/projects/", datashare, "/", gse, "/ctl_d_h2b_end-to-end_trim30_srt.bam "),
  paste0("/home/fchuffar/projects/", datashare, "/", gse, "/can_m_h2b_end-to-end_trim30_srt.bam "),
  ""
)
print(paste(cmd, args))
system2(cmd, args)
system2("cat", paste0(prefix, "_count_", mmq, ".txt.summary"))

tab = read.table("tss_count_30.txt", header=TRUE)
colnames(tab) = substr(colnames(tab), 55, 72)
head(tab)

pdf()
layout(matrix(1:2,1), respect=TRUE)
apply(tab[,-(1:6)], 2, function(cln) {
  plot(density(log2(cln+1)))
  plot(density(cln, from=-100, to=1000))
})
dev.off()

```







# Heatmap


```{r, label="computeMatrix tss_keys", eval=TRUE}
  dir.create("tmp", showWarnings=FALSE)

  matrix_file = paste0("tmp/matrix_", prefix, ".txt.gz")

  # score_filenames
  score_filenames = c(
    paste0("/home/fchuffar/projects/", datashare, "/", gse, "/ctl_d_h2b_end-to-end_trim30_srt_PE_30_4_RPKM.bw"),
    paste0("/home/fchuffar/projects/", datashare, "/", gse, "/can_m_h2b_end-to-end_trim30_srt_PE_30_4_RPKM.bw"),
    NULL
  )
  brsl = arsl = 5000
  bin_size = 100
  matrix_file = deeptoolsr::dt_compute_matrix(regions_filename=regions_filenames, score_filename=score_filenames, out_filename=matrix_file,
    blacklist_filename = blacklist_filename,
    bin_size = bin_size,
    before_region_start_length = brsl,
    after_region_start_length = arsl,
    FORCE_EXEC=TRUE,
    number_of_processors=32,
    sort_regions="descend"
  )
  ```


```{r eval=TRUE, label="sort_heatmap"}
source("common.R")
ordered_wig_signal = fast_matrix_file(matrix_file, GRP=TRUE)
hm_out_filename = paste0("tmp/hm_", prefix, ".fasted.png")
hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=paste0(matrix_file, ".fast.txt.gz"), out_filename=hm_out_filename, FORCE_EXEC=FALSE, color_map="YlOrRd", sort_regions="descend")
```

```{r, echo=FALSE, eval=TRUE, out.width="100%", results="markup"}
knitr::include_graphics(hm_out_filename)
```








# Session Information

```{r results="verbatim"}
sessionInfo()
date()
```



























