---
title: "Peak Calling with `Danpos2`"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("config")
```

```{r eval=FALSE, echo=FALSE}
# peak calling with danpos2
__conda_setup="$('/summer/epistorage/miniconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
   eval "$__conda_setup"
else
   if [ -f "/summer/epistorage/miniconda3/etc/profile.d/conda.sh" ]; then
       . "/summer/epistorage/miniconda3/etc/profile.d/conda.sh"
   else
       export PATH="/summer/epistorage/miniconda3/bin:$PATH"
   fi
fi
unset __conda_setup

conda activate danpos_env


cd ~/projects/datashare_epistorage/chip_hira_ssrp1
# ls -lha *_end-to-end_trim30_srt.bam
# samtools view -bq 30 HIRA__WT_rep1_end-to-end_trim30_srt.bam > HIRA__WT_rep1_end-to-end_trim30_srt_mmq30.bam ; samtools index HIRA__WT_rep1_end-to-end_trim30_srt_mmq30.bam

# Danpos dpos Simplicate bg sub
rm -Rf HIRA__KO_rep1; mkdir HIRA__KO_rep1; ln -s ../HIRA__KO_rep1_end-to-end_trim30_srt_mmq30.bam HIRA__KO_rep1/.; ls -lha HIRA__KO_rep1  
rm -Rf HIRA__KO_rep2; mkdir HIRA__KO_rep2; ln -s ../HIRA__KO_rep2_end-to-end_trim30_srt_mmq30.bam HIRA__KO_rep2/.; ls -lha HIRA__KO_rep2  
rm -Rf HIRA__WT_rep1; mkdir HIRA__WT_rep1; ln -s ../HIRA__WT_rep1_end-to-end_trim30_srt_mmq30.bam HIRA__WT_rep1/.; ls -lha HIRA__WT_rep1  
rm -Rf HIRA__WT_rep2; mkdir HIRA__WT_rep2; ln -s ../HIRA__WT_rep2_end-to-end_trim30_srt_mmq30.bam HIRA__WT_rep2/.; ls -lha HIRA__WT_rep2  
rm -Rf Input_KO_rep1; mkdir Input_KO_rep1; ln -s ../Input_KO_rep1_end-to-end_trim30_srt_mmq30.bam Input_KO_rep1/.; ls -lha Input_KO_rep1  
rm -Rf Input_KO_rep2; mkdir Input_KO_rep2; ln -s ../Input_KO_rep2_end-to-end_trim30_srt_mmq30.bam Input_KO_rep2/.; ls -lha Input_KO_rep2  
rm -Rf Input_WT_rep1; mkdir Input_WT_rep1; ln -s ../Input_WT_rep1_end-to-end_trim30_srt_mmq30.bam Input_WT_rep1/.; ls -lha Input_WT_rep1  
rm -Rf Input_WT_rep2; mkdir Input_WT_rep2; ln -s ../Input_WT_rep2_end-to-end_trim30_srt_mmq30.bam Input_WT_rep2/.; ls -lha Input_WT_rep2  
rm -Rf SSRP1_KO_rep1; mkdir SSRP1_KO_rep1; ln -s ../SSRP1_KO_rep1_end-to-end_trim30_srt_mmq30.bam SSRP1_KO_rep1/.; ls -lha SSRP1_KO_rep1  
rm -Rf SSRP1_KO_rep2; mkdir SSRP1_KO_rep2; ln -s ../SSRP1_KO_rep2_end-to-end_trim30_srt_mmq30.bam SSRP1_KO_rep2/.; ls -lha SSRP1_KO_rep2  
rm -Rf SSRP1_WT_rep1; mkdir SSRP1_WT_rep1; ln -s ../SSRP1_WT_rep1_end-to-end_trim30_srt_mmq30.bam SSRP1_WT_rep1/.; ls -lha SSRP1_WT_rep1  
rm -Rf SSRP1_WT_rep2; mkdir SSRP1_WT_rep2; ln -s ../SSRP1_WT_rep2_end-to-end_trim30_srt_mmq30.bam SSRP1_WT_rep2/.; ls -lha SSRP1_WT_rep2  

rm -Rf Input_WT; mkdir Input_WT; ln -s ../Input_WT_rep1_end-to-end_trim30_srt_mmq30.bam ../Input_WT_rep2_end-to-end_trim30_srt_mmq30.bam Input_WT/.; ls -lha Input_WT
rm -Rf Input_KO; mkdir Input_KO; ln -s ../Input_KO_rep1_end-to-end_trim30_srt_mmq30.bam ../Input_KO_rep2_end-to-end_trim30_srt_mmq30.bam Input_KO/.; ls -lha Input_KO

rm -Rf Input_rep1; mkdir Input_rep1; ln -s ../Input_WT_rep1_end-to-end_trim30_srt_mmq30.bam ../Input_KO_rep1_end-to-end_trim30_srt_mmq30.bam Input_rep1/.; ls -lha Input_rep1
rm -Rf Input_rep2; mkdir Input_rep2; ln -s ../Input_WT_rep2_end-to-end_trim30_srt_mmq30.bam ../Input_KO_rep2_end-to-end_trim30_srt_mmq30.bam Input_rep2/.; ls -lha Input_rep2
rm -Rf Input; mkdir Input; ln -s ../Input_WT_rep1_end-to-end_trim30_srt_mmq30.bam ../Input_KO_rep1_end-to-end_trim30_srt_mmq30.bam ../Input_WT_rep2_end-to-end_trim30_srt_mmq30.bam ../Input_KO_rep2_end-to-end_trim30_srt_mmq30.bam Input/.; ls -lha Input

rm -Rf GSM1723636; mkdir GSM1723636; ln -s ~/projects/datashare/GSE70312/GSM1723636_end-to-end_trim30_srt_mmq30.bam GSM1723636/.; ls -lha GSM1723636  



a=10
q=20
z=20
jw=40
jd=100

danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_HIRA__Input_WT_rep1 HIRA__WT_rep1 -b Input_WT_rep1 > danpos_out_HIRA__Input_WT_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_HIRA__Input_WT_rep2 HIRA__WT_rep2 -b Input_WT_rep2 > danpos_out_HIRA__Input_WT_rep2.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_SSRP1_Input_WT_rep1 SSRP1_WT_rep1 -b Input_WT_rep1 > danpos_out_SSRP1_Input_WT_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_SSRP1_Input_WT_rep2 SSRP1_WT_rep2 -b Input_WT_rep2 > danpos_out_SSRP1_Input_WT_rep2.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_HIRA__Input_KO_rep1 HIRA__KO_rep1 -b Input_KO_rep1 > danpos_out_HIRA__Input_KO_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_HIRA__Input_KO_rep2 HIRA__KO_rep2 -b Input_KO_rep2 > danpos_out_HIRA__Input_KO_rep2.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_SSRP1_Input_KO_rep1 SSRP1_KO_rep1 -b Input_KO_rep1 > danpos_out_SSRP1_Input_KO_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_SSRP1_Input_KO_rep2 SSRP1_KO_rep2 -b Input_KO_rep2 > danpos_out_SSRP1_Input_KO_rep2.log &

danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_WT_rep1 Input_WT_rep1 > danpos_out_Input_WT_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_WT_rep2 Input_WT_rep2 > danpos_out_Input_WT_rep2.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_KO_rep1 Input_KO_rep1 > danpos_out_Input_KO_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_KO_rep2 Input_KO_rep2 > danpos_out_Input_KO_rep2.log &

danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_WT Input_WT > danpos_out_Input_WT.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_KO Input_KO > danpos_out_Input_KO.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_rep1 Input_rep1 > danpos_out_Input_rep1.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input_rep2 Input_rep2 > danpos_out_Input_rep2.log &
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_Input Input > danpos_out_Input.log &

danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_GSM1723636 GSM1723636 > danpos_out_GSM1723636.log &


tail danpos_out_HIRA__Input_WT_rep1.log
tail danpos_out_HIRA__Input_WT_rep2.log
tail danpos_out_SSRP1_Input_WT_rep1.log
tail danpos_out_SSRP1_Input_WT_rep2.log
tail danpos_out_HIRA__Input_KO_rep1.log
tail danpos_out_HIRA__Input_KO_rep2.log
tail danpos_out_SSRP1_Input_KO_rep1.log
tail danpos_out_SSRP1_Input_KO_rep2.log

tail danpos_out_Input_WT.log 
tail danpos_out_Input_KO.log 
tail danpos_out_Input_rep1.log 
tail danpos_out_Input_rep2.log 
tail danpos_out_Input.log 

tail danpos_out_GSM1723636.log 



cd ~/projects/atad2/results/chip_hira_ssrp1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_HIRA__Input_WT_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_HIRA__Input_WT_rep2
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_SSRP1_Input_WT_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_SSRP1_Input_WT_rep2
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_HIRA__Input_KO_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_HIRA__Input_KO_rep2
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_SSRP1_Input_KO_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_SSRP1_Input_KO_rep2

ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_WT_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_WT_rep2
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_KO_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_KO_rep2

ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_WT
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_KO
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_rep1
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input_rep2
ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_Input

ln -s ~/projects/datashare_epistorage/chip_hira_ssrp1/danpos_out_GSM1723636


abs_thresh_smt = 96; danpos_out_dir = "danpos_out_HIRA__Input_WT_rep1"; prefix = "HIRA__WT_rep1.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_SSRP1_Input_WT_rep1"; prefix = "SSRP1_WT_rep1.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_HIRA__Input_KO_rep1"; prefix = "HIRA__KO_rep1.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_SSRP1_Input_KO_rep1"; prefix = "SSRP1_KO_rep1.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_HIRA__Input_WT_rep2"; prefix = "HIRA__WT_rep2.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_SSRP1_Input_WT_rep2"; prefix = "SSRP1_WT_rep2.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_HIRA__Input_KO_rep2"; prefix = "HIRA__KO_rep2.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_SSRP1_Input_KO_rep2"; prefix = "SSRP1_KO_rep2.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))

abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_WT_rep1"; prefix = "Input_WT_rep1";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_WT_rep2"; prefix = "Input_WT_rep2";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_KO_rep1"; prefix = "Input_KO_rep1";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_KO_rep2"; prefix = "Input_KO_rep2";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_WT";      prefix = "Input_WT.Fnor";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_KO";      prefix = "Input_KO.Fnor";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_rep1";    prefix = "Input_rep1.Fnor"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input_rep2";    prefix = "Input_rep2.Fnor"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))
abs_thresh_smt = 96; danpos_out_dir = "danpos_out_Input";         prefix = "Input.Fnor";      algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))

abs_thresh_smt = 96; danpos_out_dir = "danpos_out_GSM1723636"; prefix = "GSM1723636";   algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, "_", abs_thresh_smt, ".html"))













# "/home/fchuffar/projects/datashare/GSE70312/GSM1723635_end-to-end_trim30_srt_PE_30_4_RPKM.bw",
# "/home/fchuffar/projects/datashare/GSE70312/GSM1723636_end-to-end_trim30_srt_PE_30_4_RPKM.bw",


cd ~/projects/datashare/GSE70312

rm -Rf GSM1723635; mkdir GSM1723635; ln -s ../GSM1723635_end-to-end_trim30_srt_mmq30.bam GSM1723635/.; ls -lha GSM1723635  
rm -Rf GSM1723636; mkdir GSM1723636; ln -s ../GSM1723636_end-to-end_trim30_srt_mmq30.bam GSM1723636/.; ls -lha GSM1723636  

a=10
q=20
z=20
jw=40
jd=100
danpos.py dpos -m 1 -e 1 -jw ${jw} -jd ${jd} -a ${a} -z ${z} -q ${q} -o danpos_out_GSM1723635_GSM1723636 GSM1723635 -b GSM1723636 > danpos_out_GSM1723635_GSM1723636.log &
tail -f danpos_out_GSM1723635_GSM1723636.log

danpos.py dpeak -m 1 -o danpos_out_GSM1723635_GSM1723636 GSM1723635 -b GSM1723636 > danpos_out_GSM1723635_GSM1723636.dpeak.log &
tail -f danpos_out_GSM1723635_GSM1723636.dpeak.log


cd ~/projects/atad2/results/chip_hira_ssrp1
ln -s ~/projects/datashare/GSE70312/danpos_out_GSM1723635_GSM1723636

danpos_out_dir = "danpos_out_GSM1723635_GSM1723636"; prefix = "GSM1723635.bgsub"; algo="peaks";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, ".html"))
danpos_out_dir = "danpos_out_GSM1723635_GSM1723636"; prefix = "GSM1723635.bgsub"; algo="positions";rmarkdown::render("04_danpos_call.Rmd", output_file=paste0("04_danpos_call_", danpos_out_dir, "_", prefix, "_", algo, ".html"))

```











```{r eval=TRUE, label="params"}
quantile_thresh_smt = 0.05
quantile_thresh_fuzziness = 0.95
thresh_len = c(90, 220)

if (!exists("algo")) {
  algo="positions"
}

if (!exists("abs_thresh_smt")) {
  abs_thresh_smt = 128
}

if (!exists("prefix")) {
  prefix = "Input_rep1"
}

if (!exists("danpos_out_dir")) {
 # a=10 ; q=20 ; z=20 ; jw=40 ; jd=100 ; danpos_out_dir = paste0("danpos_out_custom_e1_jw", jw, "_jd", jd, "_a", a, "_z", z, "_q", q, "")
  danpos_out_dir = "danpos_out_Input_rep1"
}
```


```{r eval=TRUE, label="loading annotations"}
if (!exists("rmsk_raw")) {
  rmsk_raw = read.table("~/projects/small_structs/data/rmsk.mm10.bed", skip=1, stringsAsFactors=FALSE)
}
rmsk = rmsk_raw

if (!exists("annotations")) {
  # annots = c("mm10_lncrna_gencode")
  # annots = c('mm10_cpgs', 'mm10_basicgenes', "mm10_lncrna_gencode", "mm10_enhancers_fantom")
  # annots = c('mm10_cpgs', 'mm10_basicgenes', "mm10_enhancers_fantom")
  # annots = c( 'mm10_basicgenes', 'mm10_cpgs')
  write.table(rmsk[rmsk[,4]=="GSAT_MM",]  , file="gsat.mm10.bed"  , sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[rmsk[,4]=="SYNREP_MM",], file="synrep.mm10.bed", sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)

  annotatr::read_annotations(con="gsat.mm10.bed"  , genome = "mm10", name = "gsat"  , format = "bed")
  annotatr::read_annotations(con="synrep.mm10.bed", genome = "mm10", name = "synrep", format = "bed")
  annots = c( "mm10_basicgenes", "mm10_cpgs", "mm10_custom_gsat", "mm10_custom_synrep")
  annotations = annotatr::build_annotations(genome = 'mm10', annotations = annots)  
}
```



```{r eval=TRUE, label="danpos_out"}
source("common.R")

print("Reading danpos output...")
danpos_out = mget_danpos_out(prefix, danpos_out_dir, algo=algo)
head(danpos_out)
print("Annotating danpos output...")
danpos_out_annot = mget_danpos_out_annot(prefix, danpos_out_dir, algo=algo)
head(danpos_out_annot)
table(danpos_out_annot$annot.type)
# stope("EFN")

# danpos_out[danpos_out$smt_value==16 & danpos_out$fuzziness_score <30, ]
# head(danpos_out[danpos_out$smt_value==16 & danpos_out$fuzziness_score <35, ])


if (!exists("chr_sizes")) {
  print("Reading chr_sizes...")
  chr_sizes = read.table("~/projects/datashare/genomes/Mus_musculus/UCSC/mm10/mm10.chrom.sizes", stringsAsFactors=FALSE)
  rownames(chr_sizes) = chr_sizes[,1]
  head(chr_sizes)
  dim(chr_sizes)
}
genome_size = sum(chr_sizes[names(table(danpos_out$chr)),2])


# FILTERING DANPOS OUT
# danpos_out[idx, ][!duplicated()]
head(danpos_out)
dim(danpos_out)
print(paste0(signif(nrow(danpos_out) * 140 / 1000000, 3), " Mb (over ", signif(genome_size / 1000000, 3), " Mb)"))
print(table(danpos_out$chr))

sum(danpos_out$len)

print(paste0(signif(100 * sum(danpos_out$len) / genome_size,3), "% of the genome"))

```

# Params

Peak calling done on `r prefix` with danpos params `r danpos_out_dir`, with algorithm `r algo`.

```{r stats, results="verbatim"}
print(paste0(signif(100 * sum(danpos_out$len) / genome_size,3), "% of the genome"))
```



```{r results="verbatim", eval=FALSE}
print(table(danpos_out_annot$annot.type))
print(round(table(danpos_out_annot$annot.type)/nrow(danpos_out)*100,2))
```



















# Filtering features

```{r, label="filtering", results="verbatim"}
if (algo=="peaks") {
  print(paste0("algo used:", algo))
  danpos_out$smt_value = danpos_out$total_signal
  danpos_out$fuzziness_score = danpos_out$width_above_cutoff
  danpos_out$smt_pos = danpos_out$center
  # thresh_fuzziness_score = 5000
  # tmp_sub_danpos_out = danpos_out[danpos_out$fuzziness_score <= thresh_fuzziness_score & danpos_out$smt_value >= thresh_smt & !rownames(danpos_out) %in% danpos_out_annot[danpos_out_annot$annot.type %in% c("mm10_custom_gsat","mm10_custom_synrep"),]$name,]
} else {
  print(paste0("algo used:", algo))
}

plot_danpos_out_stat = function(tmp_sub_danpos_out, genome_size) {
  layout(matrix(1:3,1), respect=TRUE)
  barplot(table(tmp_sub_danpos_out$len), las=2, main=paste0("distribution of ", nrow(tmp_sub_danpos_out), " regions lengths"))
  boxplot(log2(tmp_sub_danpos_out$smt_value+1)~tmp_sub_danpos_out$len, las=2, main="log(smt_value)~len")
  boxplot(tmp_sub_danpos_out$fuzziness_score~tmp_sub_danpos_out$len, las=2, main="fuzziness~len")
  print(paste0(signif(100 * sum(tmp_sub_danpos_out$len) / genome_size,3), "% of the genome"))  
  dim(tmp_sub_danpos_out)
}

print("No filter")
tmp_sub_danpos_out = danpos_out
tmp_sub_danpos_out_annot = danpos_out_annot
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)

print("filtering according to minor and major sat")
tmp_sub_danpos_out = tmp_sub_danpos_out[
  !rownames(tmp_sub_danpos_out) %in% tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$annot.type %in% c("mm10_custom_gsat","mm10_custom_synrep"),]$name,
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)

print("filtering according to the length of the features")
tmp_sub_danpos_out = tmp_sub_danpos_out[
  tmp_sub_danpos_out$len >= thresh_len[1] &
  tmp_sub_danpos_out$len <= thresh_len[2],
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)

print("filtering Mito")
tmp_sub_danpos_out = tmp_sub_danpos_out[
  !tmp_sub_danpos_out$chr %in% "chrM", 
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)


print("filtering according to the level of the signal (quantile)")
dict_thresh_smt = sapply(as.character(sort(unique(tmp_sub_danpos_out$len))), function(len) {
  quantile(tmp_sub_danpos_out[tmp_sub_danpos_out$len==len,]$smt_value, probs=quantile_thresh_smt)[[1]]
})
tmp_sub_danpos_out = tmp_sub_danpos_out[
  tmp_sub_danpos_out$smt_value > dict_thresh_smt[as.character(tmp_sub_danpos_out$len)],
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)


print("filtering according to the level of the signal (quantile)")
dict_thresh_smt = sapply(as.character(sort(unique(tmp_sub_danpos_out$len))), function(len) {
  quantile(tmp_sub_danpos_out[tmp_sub_danpos_out$len==len,]$smt_value, probs=0.999)[[1]]
})
tmp_sub_danpos_out = tmp_sub_danpos_out[
  tmp_sub_danpos_out$smt_value < dict_thresh_smt[as.character(tmp_sub_danpos_out$len)],
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)



print(paste0("filtering according to the level of the signal (absolute threshold ", abs_thresh_smt, ")"))
tmp_sub_danpos_out = tmp_sub_danpos_out[
  tmp_sub_danpos_out$smt_value > abs_thresh_smt,
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)

print("filtering according to the fuzziness of the signal")
dict_thresh_fuzziness = sapply(as.character(sort(unique(tmp_sub_danpos_out$len))), function(len) {
  quantile(tmp_sub_danpos_out[tmp_sub_danpos_out$len==len,]$fuzziness_score, probs=quantile_thresh_fuzziness)[[1]]
})
tmp_sub_danpos_out = tmp_sub_danpos_out[
  tmp_sub_danpos_out$fuzziness_score < dict_thresh_fuzziness[as.character(tmp_sub_danpos_out$len)],
]
tmp_sub_danpos_out_annot = tmp_sub_danpos_out_annot[tmp_sub_danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
plot_danpos_out_stat(tmp_sub_danpos_out, genome_size)
```

# Plotting fuzziness and occupancy



```{r, label="fuzziness/occupancy plots"}
print("categorize features according to fuzziness...")
if (nrow(tmp_sub_danpos_out) < 50000) {
  q = unique(quantile(tmp_sub_danpos_out$fuzziness_score, probs=seq(0,1,length=6)))
} else if (nrow(tmp_sub_danpos_out) < 500000) {
  q = unique(quantile(tmp_sub_danpos_out$fuzziness_score, probs=c(0, 0.05, seq(0,1,length=6)[-1])))  
} else {
  q = unique(quantile(tmp_sub_danpos_out$fuzziness_score, probs=c(0, 0.005, 0.05, seq(0,1,length=6)[-1])))  
}
tmp_sub_danpos_out$grp = as.numeric(cut(tmp_sub_danpos_out$fuzziness_score, q, include.lowest=TRUE))
print(table(tmp_sub_danpos_out$grp))

# # Export gene lists for Amigo
# dir.create("tmp", recursive=TRUE, showWarnings=FALSE)
# for (i in unique(tmp_sub_danpos_out$grp)) {
#   genesqi_file = paste0("tmp/genesq", i, "_", prefix, ".txt")
#   genesqi = unique(danpos_out_annot[danpos_out_annot$annot.type %in% "mm10_genes_promoters" & danpos_out_annot$name%in% rownames(tmp_sub_danpos_out)[tmp_sub_danpos_out$grp%in%i],]$annot.symbol)
#   write.table(genesqi, genesqi_file, row.names=FALSE, quote=FALSE, col.names=FALSE)
# }


layout(matrix(1:6, 2), respect=TRUE)

plot( density(log2(        danpos_out$smt_value + 1)), main="raw log2(occupancy+1) distribution")
lines(density(log2(tmp_sub_danpos_out$smt_value + 1)), col=adjustcolor(2, alpha.f=.5))
# abline(v=log2(thresh_smt + 1), col="red", lty=2)

plot( density(        danpos_out$fuzziness_score), main="raw fuzziness distribution")
lines(density(tmp_sub_danpos_out$fuzziness_score), col=adjustcolor(2, alpha.f=.5))
abline(v=q, lty=3, col="grey")
abline(v=q[2], lty=2, col="blue")

plot(density(tmp_sub_danpos_out$fuzziness_score), main="filtered fuzziness distribution")
abline(v=q, lty=3, col="grey")
abline(v=q[2], lty=2, col="blue")

smoothScatter(
  tmp_sub_danpos_out$fuzziness_score, log2(tmp_sub_danpos_out$smt_value + 1),
  ylab="log2(occupancy+1) (a.u)", 
  ylim=range(log2(tmp_sub_danpos_out$smt_value + 1)), 
  col=1, pch=1
)
abline(v=q, lty=3, col="grey")
abline(v=q[2], lty=2, col="blue")

barplot(round(table(tmp_sub_danpos_out_annot$annot.type)/nrow(tmp_sub_danpos_out)*100,2), las=2)
tmp_d = density(log2(tmp_sub_danpos_out$smt_value + 1))

plot(tmp_d$y, tmp_d$x, type="l", xlab="Density", ylab="",
ylim=range(log2(tmp_sub_danpos_out$smt_value + 1))
)

foo = tmp_sub_danpos_out[tmp_sub_danpos_out$smt_value==min(tmp_sub_danpos_out[tmp_sub_danpos_out$grp==1,]$smt_value) & tmp_sub_danpos_out$grp==1,]
head(foo)

foo = tmp_sub_danpos_out[tmp_sub_danpos_out$smt_value==64 & tmp_sub_danpos_out$grp==1,]
head(foo)

# rsync -auvP cargo:~/projects/datashare_epistorage/chip_hira_ssrp1/Input_WT_rep1_end-to-end_trim30_srt_mmq30.bam* ~/projects/datashare_epistorage/chip_hira_ssrp1/.

```



























```{r eval=TRUE, label="regions_filenames"}
nb_rnd_feat = min(2500, min(table(tmp_sub_danpos_out$grp)))
annot_stats = lapply(1:5, function(seed) {
  lapply(sort(unique(tmp_sub_danpos_out$grp)), function(grp) {
    regions_filename = paste0("danpos_", prefix, "_", nb_rnd_feat, "_grp", grp, "_rnd", seed,".bed")
    print(paste0("Filtering ", nb_rnd_feat, " rnd nucleosomes in ", regions_filename, "..."))
    feat = tmp_sub_danpos_out[tmp_sub_danpos_out$grp==grp,]
    set.seed(seed)
    feat = feat[sample(1:nrow(feat), nb_rnd_feat),]
    feat[,2] = feat$smt_pos
    feat[,3] = feat[,2] + 1
    feat[,4] = rownames(feat)
    feat[,5] = grp
    feat[,6] = "+"
    # print(head(feat))
    dim(feat)
    write.table(feat, file=regions_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
    regions_filename 
    ret = table(danpos_out_annot[danpos_out_annot$name %in% rownames(feat),]$annot.type)
    ret = c(ret, grp=grp, rep=seed)
    return(ret)
  })
}) 
annot_stats = unlist(annot_stats, recursive=FALSE)
annot_stats = do.call(rbind,annot_stats)

regions_filenames_all = lapply(1:5, function(seed) {
  regions_filenames = sapply(sort(unique(tmp_sub_danpos_out$grp)), function(grp) {
    regions_filename = paste0("danpos_", prefix, "_", nb_rnd_feat, "_grp", grp, "_rnd", seed,".bed")
    regions_filename 
  })
  regions_filenames
}) 
regions_filenames = regions_filenames_all[[1]]
regions_filenames_all = unlist(regions_filenames_all)
```

# Annotations


```{r results="verbatim"}
print(table(tmp_sub_danpos_out$grp))
```

```{r label="boxplot annot"}
# layout(matrix(1:18, 3), respect=TRUE)
# layout(matrix(1:2, 1), respect=TRUE)
layout(matrix(1:6, 2), respect=TRUE)
sapply(rev(colnames(annot_stats))[-(1:2)],function(cn) {
  boxplot(annot_stats[,cn]~annot_stats[,"grp"], main=cn)
})

```






















# Heatmap



```{r label="blacklist"}
options(scipen=999)

# blacklist = rmsk
blacklist = rmsk[rmsk[,4] %in% c("GSAT_MM","SYNREP_MM"),]#, "(CTGTG)n"),]#
blacklist = rbind(blacklist, list("chr19", 61266781, 61266782, "chr19:61266781-61266782", 1404, "+"))
blacklist = rbind(blacklist, list("chr11", 3126500, 3200500, "Sfi1", 300, "-"))
# sat2_feat_saf = rbind(sat2_feat_saf, list("Sfi1", "chr11", 3126500, 3200500, "-"))
print(head(blacklist))
dim(blacklist)
blacklist_filename = "blacklist_file.bed"
write.table(blacklist, file=blacklist_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)    
```





```{r eval=TRUE, label="matrix_file"}
matrix_file = paste0("tmp/matrix_", prefix, ".txt.gz")      

# score_filenames
score_filenames = c(
  # "/home/fchuffar/projects/datashare/GSE70312/GSM1723635_end-to-end_trim30_srt_PE_30_4_RPKM.bw",
  # "/home/fchuffar/projects/datashare/GSE70312/GSM1723636_end-to-end_trim30_srt_PE_30_4_RPKM.bw",
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_WT_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_KO_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_WT_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__WT_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_KO_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__KO_rep1_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_WT_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_KO_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_WT_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__WT_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_KO_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__KO_rep2_end-to-end_trim30_srt_PE_30_4_RPKM.bw" ,
  NULL
)
brsl = arsl = 2000
bin_size = 10
matrix_file = deeptoolsr::dt_compute_matrix(regions_filename=regions_filenames, score_filename=score_filenames, out_filename=matrix_file,
  blacklist_filename = blacklist_filename,
  bin_size = bin_size,
  before_region_start_length = brsl, 
  after_region_start_length = arsl,
  FORCE_EXEC=TRUE, 
  number_of_processors=32
)
# hm_out_filename = paste0("tmp/hm_", prefix, ".raw.png")
# hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=matrix_file, out_filename=hm_out_filename, FORCE_EXEC=TRUE, color_map="YlOrRd")
```

```{r eval=TRUE, label="sort_heatmap"}
source("common.R")
ordered_wig_signal = fast_matrix_file(matrix_file, GRP=TRUE)
hm_out_filename = paste0("tmp/hm_", prefix, ".fasted.png")
hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=paste0(matrix_file, ".fast.txt.gz"), out_filename=hm_out_filename, FORCE_EXEC=FALSE, color_map="YlOrRd", sort_regions="descend")
```

```{r, echo=FALSE, eval=TRUE, out.width="100%", results="markup"}
knitr::include_graphics(hm_out_filename)
```


```{r eval=TRUE, label="heatmap"}
# sort_matrix_file = function(matrix_file, GRP=FALSE) {
#   raw_wig_signal = read.table(gzfile(matrix_file), skip=1, stringsAsFactors=FALSE)
#   dim(raw_wig_signal)
#   # plot(apply(raw_wig_signal[,-(1:6)], 2, mean))
#   # abline(v=c(131,150), col=2)
#
#   # filter_according_to_column_nb = 5
#   # idx = ((filter_according_to_column_nb-1) * (brsl + arsl)/bin_size + 1):(filter_according_to_column_nb*(brsl + arsl)/bin_size)
#   # q = quantile(apply(as.matrix(raw_wig_signal[1:1000,-(1:6)][,idx]), 1, mean), probs=0.990)
#   # idx = which(as.vector(apply(as.matrix(raw_wig_signal[1:1000,-(1:6)][,idx]), 1,mean) > q))
#   # # raw_wig_signal[idx,1:6][]
#   # raw_wig_signal[idx,-(1:6)] = 0
#
#   sort_according_to_column_nb = 3
#   idx = ((sort_according_to_column_nb-1) * (brsl + arsl)/bin_size + 1):(sort_according_to_column_nb*(brsl + arsl)/bin_size)
#   score = apply(raw_wig_signal[,-(1:6)][,idx], 1, function(l){
#     # l = raw_wig_signal[7427,-(1:6)][,idx]
#     which(l%in%max(l, ra.rm=TRUE))[1]
#   })
#   mat_centred_on_peak = raw_wig_signal[,1:6]
#   mat_centred_on_peak[,2] = mat_centred_on_peak[,2] - brsl + score*bin_size
#   mat_centred_on_peak[,3] = mat_centred_on_peak[,2] + 1
#
#
#   grps = raw_wig_signal[,5]
#   system2(command="zcat", args=paste0(matrix_file, " | head -n1 ",
#     "| sed -e 's/_end-to-end_trim30_srt_PE_30_4_RPKM//g' ",
#     "| sed -e 's/_end-to-end_trim30_srt_30_4_RPKM//g' ",
#     "| sed -e 's/_end-to-end_trim30_fsmin125_fsmax175_srt_30_4_RPKM/_nuc/g' ",
#     "| sed -e 's/_end-to-end_trim30_fsmin025_fsmax075_srt_30_4_RPKM/_ss/g' ",
#     "| sed -e 's/_srt_30_4_RPKM//g' ",
#     "| sed -e 's/SRR1019858/GSM1253297/g' ",
#     "| sed -e 's/SRR1019859/GSM1253298/g' ",
#     "| sed -e 's/SRR1019860/GSM1253299/g' ",
#     "| sed -e 's/SRR1019861/GSM1253300/g' ",
#     "| sed -e 's/SRR1019862/GSM1253301/g' ",
#     "| sed -e 's/SRR1019863/GSM1253302/g' ",
#     "| sed -e 's/S_bn3/S_nuc_4/g' ",
#     "| sed -e 's/S_bn1/S_nuc_12/g' ",
#     "| sed -e 's/S_bn6/S_nuc_12/g' ",
#     "| sed -e 's/S_bn5/S_sms_4/g' ",
#     "| sed -e 's/S_bn2/S_sms_12/g' ",
#     "| sed -e 's/S_pp2/S_WT_04/g' ",
#     "| sed -e 's/S_pp4/S_WT_08/g' ",
#     "| sed -e 's/S_pp6/S_WT_12/g' ",
#     "| sed -e 's/S_nn2/S_KO_04/g' ",
#     "| sed -e 's/S_nn4/S_KO_08/g' ",
#     "| sed -e 's/S_nn6/S_KO_12/g' ",
#     "> ", matrix_file, ".sort.txt"))
#   if (GRP) {
#     o = order(grps, score)
#   } else {
#     o = order(score)
#   }
#
#
#   # write.table(raw_wig_signal[o,], paste0(matrix_file, ".sort.txt"), quote=FALSE, col.names=FALSE, row.names=FALSE, append=TRUE, sep="\t")
#   write.table(raw_wig_signal[,], paste0(matrix_file, ".sort.txt"), quote=FALSE, col.names=FALSE, row.names=FALSE, append=TRUE, sep="\t", na="nan")
#   system2(command="rm", args=paste0(matrix_file, ".sort.txt.gz"))
#   system2(command="gzip", args=paste0(matrix_file, ".sort.txt"))
#   ret = list(ordered_wig_signal=raw_wig_signal[o,] , centred_on_peak_bed=mat_centred_on_peak[o,])
#   return(ret)
# }
#
#
# ordered_wig_signal = sort_matrix_file(matrix_file, GRP=TRUE)[[1]]
#
# # ordered_wig_signal_file = paste0("tmp/ordered_wig_signal_", prefix, "_", thresh_smt, ".txt.gz")
# # saveRDS(ordered_wig_signal, ordered_wig_signal_file)
# #
# # brsl = arsl = 200
# # bin_size = 10
# # prefix = "S_pp2_end-to-end_trim30_fsmin125_fsmax175_srt_mmq30"
# # thresh_smt = 96
# # data = readRDS(paste0("tmp/ordered_wig_signal_", prefix, "_", thresh_smt, ".txt.gz"))
# # dim(data)
# #
# # filter_according_to_column_nb = c(15)
# # idx = unlist(lapply(filter_according_to_column_nb, function(i) {((i-1) * (brsl + arsl)/bin_size + 1):(i*(brsl + arsl)/bin_size)}))
# # head(data[,c(1:6, idx + 7)])
# # d = log2(as.matrix(data[,-(1:6)]) + 1)
# # rownames(d) = data[,4]
# # d = d[1:5000,]
# # o = order(apply(d[,idx[11:30]], 1, mean))
# # d = d[o,]
# # plot(density(d))
# # idx = rownames(d[1:1000,])
# #
# # a = danpos_out_annot[danpos_out_annot$name %in% idx,]$annot.type
# # pa = prop.table(table(a))
# # b = danpos_out_annot[danpos_out_annot$name %in%  rownames(d),]$annot.type
# # pb = prop.table(table(b))
# #
# # genes = unique(danpos_out_annot[danpos_out_annot$annot.type %in% "mm10_genes_promoters" & danpos_out_annot$name%in% idx,]$annot.symbol)
# # write.table(genes, "genes.txt"    , row.names=FALSE, quote=FALSE, col.names=FALSE)
# #
# #
# # source("~/projects/epimedpipelines/results/commons.R")
# # foo = plot_expr_hm(d[1:1000,], hcmeth_cols=FALSE, hcmeth_rows=FALSE)
# # foo = plot_expr_hm(d[4500:5000,], hcmeth_cols=FALSE, hcmeth_rows=FALSE)
# #
# # genes = names(sort(table(danpos_out_annot[danpos_out_annot$name %in% idx,]$annot.symbol)))
#
#
#
# # plotHeatmap --matrixFile tmp/matrix_Input_WT_rep1.txt.gz.sort.txt.gz --outFileName tmp/hm_Input_KO_rep1.sorted.png --colorMap YlOrRd --sortRegions no
# # plotHeatmap --matrixFile tmp/matrix_Input_WT_rep1.txt.gz --outFileName tmp/hm_Input_KO_rep1.sorted.png --colorMap YlOrRd --sortRegions no
#
# hm_out_filename = paste0("tmp/hm_", prefix, ".sorted.png")
# hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=paste0(matrix_file, ".sort.txt.gz"), out_filename=hm_out_filename, FORCE_EXEC=TRUE, color_map="YlOrRd")
# ```
#
# ```{r, echo=FALSE, eval=TRUE, out.width="100%", results="markup"}
# knitr::include_graphics(hm_out_filename)
```


```{r heatmap_clust}
# hm_out_filename = paste0("tmp/hm_", prefix, ".sortedclust.png")
# hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=paste0(matrix_file, ".sort.txt.gz"), out_filename=hm_out_filename, FORCE_EXEC=TRUE, color_map="YlOrRd", sort_regions="descend", raw_args="--kmeans 4")
# ```
#
# ```{r, echo=FALSE, eval=TRUE, out.width="100%", results="markup"}
# knitr::include_graphics(hm_out_filename)
```







# Counts

```{r, label="SAF stuffs", eval=TRUE}
# if (length(grep("fsmin", prefix))!=0) {
#   min_len = as.numeric(strsplit(prefix,"fsmin|_fsmax")[[1]][2])
#   max_len = as.numeric(strsplit(prefix,"fsmax|_srt")[[1]][2])
#   tmp_sub_danpos_out = tmp_sub_danpos_out[tmp_sub_danpos_out$len > min_len & tmp_sub_danpos_out$len < max_len ,]
#   tmp_sub_danpos_out_annot = danpos_out_annot[danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
#   dim(tmp_sub_danpos_out)
# }

  # min_len =  100
  # max_len = 200
  # tmp_sub_danpos_out = tmp_sub_danpos_out[tmp_sub_danpos_out$len > min_len & tmp_sub_danpos_out$len < max_len ,]
  # tmp_sub_danpos_out_annot = danpos_out_annot[danpos_out_annot$name %in% rownames(tmp_sub_danpos_out),]
  # dim(tmp_sub_danpos_out)



# export as SAF here...
# # tmp_sub_danpos_out...
# annotatr_feat_saf = as.data.frame(annotations)[,c("type", "seqnames", "start", "end", "strand")]
# names(annotatr_feat_saf) = c("GeneID", "Chr", "Start", "End", "Strand")
# head(annotatr_feat_saf)
# table(annotatr_feat_saf$GeneID)
#
# table(annotatr_feat_saf$Chr)
# annotatr_feat_saf = annotatr_feat_saf[annotatr_feat_saf$Chr %in% names(table(annotatr_feat_saf$Chr))[1:21],]
# table(annotatr_feat_saf$Chr)

feat_saf = tmp_sub_danpos_out[,c(1,1,2,3)]
feat_saf[,1] = rownames(tmp_sub_danpos_out)
feat_saf$strand= "+"
head(feat_saf)
dim(feat_saf)
feat_saf_filename = paste0(prefix, "_", danpos_out_dir, "_feat.saf")
write.table(feat_saf, file=feat_saf_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
# [1] "S_pp2_end-to-end_trim30_fsmin125_fsmax175_srt_mmq30_danpos_out_custom_e1_jw40_jd100_a10_z20_q20_96_feat.saf"

# counts
cmd = "/summer/epistorage/miniconda3/envs/subread_env/bin/featureCounts"
count_filename = paste0(prefix, "_", danpos_out_dir, "_count.txt")
mmq = 0
fc_options = paste0("-a ", feat_saf_filename , " -F SAF --largestOverlap -Q ", mmq, " -T ", parallel::detectCores(logical=FALSE), " -o ", count_filename, " ")

bam_filenames = c(
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_WT_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_KO_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_WT_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__WT_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_KO_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__KO_rep1_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_WT_rep2_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_KO_rep2_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_WT_rep2_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__WT_rep2_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/SSRP1_KO_rep2_end-to-end_trim30.bam" ,
  "/home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/HIRA__KO_rep2_end-to-end_trim30.bam" ,
  NULL
)



# ls -lha /home/fchuffar/projects/datashare/GSE70312/GSM1723636_end-to-end_trim30.bam
# ls -lha /home/fchuffar/projects/datashare_epistorage/chip_hira_ssrp1/Input_*_end-to-end_trim30.bam


bams = bam_filenames
args = paste0(fc_options, paste(bam_filenames, collapse= " "))
print(paste(cmd, args))
system2(cmd, args)      
```








```{r label="normalization"}
normalize_counts = function (count_filename) {
  # raw
  foo = read.table(count_filename, header=TRUE)
  pf = foo[,1:6]
  head(pf)
  foo = foo[, -(1:6)]
  rownames(foo) = paste0(pf[,2], ":", pf[,3], "-", pf[,4])
  colnames(foo) = substr(colnames(foo), 63, 75)
  head(foo)
  dim(foo)
  d = as.matrix(foo)
  # plot(density(d))

  # RPM normalized
  d_rpm = t(t(d) / apply(d, 2, sum) * 1000000)
  d_lrpm = log2(d_rpm + 1)
  # plot(density(d_lrpm))

  # DESeq2 normalized
  countData = d
  colData = data.frame(id=colnames(countData), lines=factor(substr(colnames(d), 1, 8)))
  dds = DESeq2::DESeqDataSetFromMatrix(countData=countData, colData=colData, design= ~ lines)
  dds = DESeq2::estimateSizeFactors(dds)
  sf = dds$sizeFactor
  sf
  norm_counts = DESeq2::counts(dds, normalized=TRUE)
  d_norm = norm_counts
  d_lnorm = log2(norm_counts + 1)

  return(list(
    pf = pf,
    raw_counts = d    ,
    d_rpm   = d_rpm   ,
    d_lrpm  = d_lrpm  ,
    d_norm  = d_norm  ,
    d_lnorm = d_lnorm ,
    NULL
  ))
}
if (!exists("mnormalize_counts")) {mnormalize_counts = memoise::memoise(normalize_counts)}
```

```{r label="PCA"}
plot_pca = function(data) {
  dim(data)
  data = data[,!duplicated(apply(data, 2, paste, collapse="_"))]
  dim(data)
  data = data[,apply(data, 2, function(c){length(unique(c))})>1]
  dim(data)
  pca = prcomp(data, scale=TRUE)
  v = pca$sdev * pca$sdev
  p = v / sum(v) * 100
  layout(matrix(1:6,2, byrow=FALSE), respect=TRUE)
  barplot(p)
  # colors = rev(RColorBrewer::brewer.pal(n=11, "Paired"))
  colors = c("red", "blue")
  keys = rownames(data)
  cols = adjustcolor(as.numeric(factor(substr(rownames(data), 1, 8))), alpha.f=0.5)
  i=3
  j=2
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=cols, pch=16, cex=5)
  text(pca$x[,i], pca$x[,j], keys)
  i=1
  j=3
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=cols, pch=16, cex=5)
  text(pca$x[,i], pca$x[,j], keys)
  i=1
  j=2
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=cols, pch=16, cex=5)
  text(pca$x[,i], pca$x[,j], keys)
  i=4
  j=5
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=cols, pch=16, cex=5)
  text(pca$x[,i], pca$x[,j], keys)  
}

foo = mnormalize_counts(count_filename)

# RPM normalized
# dev.new()
d_lrpm = foo$d_lrpm
plot_pca(as.matrix(t(d_lrpm)))
plot(density(d_lrpm))
cols = as.numeric(factor(substr(colnames(d_lrpm), 1, 8)))
names(cols) = colnames(d_lrpm)
sapply(colnames(d_lrpm), function(cn) {
  print(cn)
  # cn = colnames(d_lrpm)[1]
  lines(density(d_lrpm[,cn]), col=cols[[cn]], lty=as.numeric(substr(cn, nchar(cn),nchar(cn))))
})


# DESeq2 normalized
# dev.new()
d_lnorm = foo$d_lnorm
plot_pca(as.matrix(t(d_lnorm)))
plot(density(d_lnorm))
cols = as.numeric(factor(substr(colnames(d_lnorm), 1, 8)))
names(cols) = colnames(d_lnorm)
sapply(colnames(d_lnorm), function(cn) {
  print(cn)
  # cn = colnames(d_lnorm)[1]
  lines(density(d_lnorm[,cn]), col=cols[[cn]], lty=as.numeric(substr(cn, nchar(cn),nchar(cn))))
})
```


```{r eval=FALSE}
bar = apply(d_lnorm==0, 1, all)
sum(bar)

plot(density(d_lnorm))

d = d_lnorm[sample(1:nrow(d_lnorm),1000),]
d = d_lnorm#[sample(1:nrow(d_lnorm),1000),]

head(d)
par(mar=c(5.1, 4.1, 4.1, 2.1), mgp=c(3, 1, 0), las=0)
layout(matrix(1:2,1), respect=TRUE)
plot(d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"], d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"],   xlim=c(-4,5), ylim=c(-4,5))
points(d[,"HIRA__KO_rep1"] - d[,"Input_KO_rep1"], d[,"SSRP1_KO_rep1"] - d[,"Input_KO_rep1"], xlim=c(-4,5), ylim=c(-4,5), col=2)
abline(b=1, a=0, lty=2)
plot(d[,"HIRA__KO_rep1"] - d[,"Input_KO_rep1"], d[,"SSRP1_KO_rep1"] - d[,"Input_KO_rep1"],   xlim=c(-4,5), ylim=c(-4,5))
points(d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"], d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"], xlim=c(-4,5), ylim=c(-4,5), col=2)
abline(b=1, a=0, lty=2)






d = d_lrpm
d = d_lnorm
layout(matrix(1:3,1), respect=TRUE)
plot(
  (d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"]) - (d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"]),
  (d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"]) + (d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"]),  
  xlab="lfc", ylab="log total counts", pch="."
)
abline(h=1, v=c(-1,1), col=2, lty=2)
abline(h=0, v=0, col="grey", lty=3)

plot(
  (d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"]) - (d[,"HIRA__KO_rep1"] - d[,"Input_WT_rep1"]),
  (d[,"HIRA__WT_rep1"] - d[,"Input_WT_rep1"]) + (d[,"HIRA__KO_rep1"] - d[,"Input_WT_rep1"]),  
  xlab="lfc", ylab="log total counts", pch="."
)
abline(h=1, v=c(-1,1), col=2, lty=2)
abline(h=0, v=0, col="grey", lty=3)

plot(
  (d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"]) - (d[,"SSRP1_KO_rep1"] - d[,"Input_WT_rep1"]),
  (d[,"SSRP1_WT_rep1"] - d[,"Input_WT_rep1"]) + (d[,"SSRP1_KO_rep1"] - d[,"Input_WT_rep1"]),  
  xlab="lfc", ylab="log total counts", pch="."
)
abline(h=1, v=c(-1,1), col=2, lty=2)
abline(h=0, v=0, col="grey", lty=3)




d = d_lrpm
ana_diff = epimedtools::monitored_apply(d, 1, function(l){
  # l = d[1,]
  df = data.frame(
    lcnt = c(
      l[["SSRP1_WT_rep1"]] - l[["Input_WT_rep1"]],
      l[["SSRP1_KO_rep1"]] - l[["Input_WT_rep1"]],
      l[["HIRA__WT_rep1"]] - l[["Input_WT_rep1"]],
      l[["HIRA__KO_rep1"]] - l[["Input_WT_rep1"]]
    ),
    chip = c(
      "SSRP1",
      "SSRP1",
      "HIRA_",
      "HIRA_"
    ),
    line = c(
      "WT",
      "KO",
      "WT",
      "KO"
    )
  )
  m = lm(lcnt~chip+line, df)
  a = anova(m)
  res = c(fc_chip=m$coefficients[[2]], pv_chip=a[1,5], fc_line=m$coefficients[[3]], pv_line=a[2,5])
  return(res)
})
ana_diff = t(ana_diff)
head(ana_diff)

plot(ana_diff[,1], -log10(ana_diff[,2]))
plot(ana_diff[,3], -log10(ana_diff[,4]))
plot(-log10(ana_diff[,2]), -log10(ana_diff[,4]))
plot(ana_diff[,1], ana_diff[,3])
```


```{r comb-p, eval=FALSE}
dir.create(path="combp_results", showWarnings=FALSE)
res_EWAS = ana_diff[,1:2]
res_EWAS = res_EWAS[!is.na(ana_diff[,1]),]
res_EWAS = res_EWAS[!is.na(ana_diff[,2]),]
head(res_EWAS)
tmp_split = strsplit(rownames(res_EWAS),":|-")
tmp_split = do.call(rbind, tmp_split)
head(tmp_split)
methpf = data.frame(id=rownames(res_EWAS), chr=tmp_split[,1] ,pos=as.numeric(tmp_split[,2]) ,stringsAsFactors=FALSE)
rownames(methpf) = methpf$id
methpf = methpf[,-1]
bedfile = methpf
colnames(bedfile) = c("chrom","start")
bedfile$end = bedfile$start + 1
head(bedfile)

ewasforcombp2 = merge(bedfile,res_EWAS, by=0)
ewasforcombp2 = as.data.frame(ewasforcombp2)
rownames(ewasforcombp2) = ewasforcombp2[,1]
ewasforcombp2 = ewasforcombp2[,-1]






head(tmp_sub_danpos_out)
ewasforcombp2 = tmp_sub_danpos_out[, c(1, 4, 4, 8,5)]
ewasforcombp2[,3] = ewasforcombp2[,2] + 1
ewasforcombp2[,5] = 10^-log2(ewasforcombp2[,5]+1)
ewasforcombp2 = ewasforcombp2[ewasforcombp2$grp<=2,]

colnames(ewasforcombp2) = c("chrom","start", "end", "grp", "pval")
head(ewasforcombp2)
table(ewasforcombp2[,1])





ewasforcombp2 = ewasforcombp2[order(ewasforcombp2$chrom, ewasforcombp2$start),]
head(ewasforcombp2)
dim(ewasforcombp2)

print("writing ewasforcombp...")
head(ewasforcombp2)
write.table(ewasforcombp2, file=paste0("combp_results/ewasforcombp_",prefix,".bed"), sep="\t", quote=FALSE,row.names=FALSE, col.names=TRUE)
print("done.")

# quantile(ewasforcombp2[,5])

print("launching comb-p...")
cmd ="/summer/epistorage/miniconda3/bin/python"
arg = paste0("/summer/epistorage/opt/combined-pvalues/cpv/comb-p pipeline --no-fdr -c 5 --seed 0.05 --dist 2000 -p combp_results/dmrbycombp1000_",prefix," --region-filter-p 0.05 --region-filter-n 2 combp_results/ewasforcombp_",prefix,".bed")
print(paste(cmd, arg))
system2(cmd, arg)
print("done.")
```












```

# # DESeq2
# foo = read.table(count_filename, header=TRUE)
# pf = foo[,1:6]
# head(pf)
# foo = foo[, -(1:6)]
# rownames(foo) = paste0(pf[,2], ":", pf[,3], "-", pf[,4])
# colnames(foo) = substr(colnames(foo), 63, 75)
# head(foo)
# dim(foo)
# d =
#
#
# countData = as.matrix(foo)
# head(countData)
# # countData = cnts[,7:12]
# # countData = cnts[,13:18]
# head(countData)
# tail(countData)
# dim(countData)
# colData = data.frame(id=colnames(countData), h2al2=c("WT", "WT", "WT", "KO", "KO", "KO"), time=c("T1", "T2", "T3", "T1", "T2", "T3"))
#
# [1] "Input_WT_rep1" "Input_KO_rep1" "SSRP1_WT_rep1" "HIRA__WT_rep1"
# [5] "SSRP1_KO_rep1" "HIRA__KO_rep1"
# >
#
# dds = DESeq2::DESeqDataSetFromMatrix(countData=countData, colData=colData, design= ~ h2al2 + time)
# dds = DESeq2::estimateSizeFactors(dds)
#
# sf = dds$sizeFactor
# sf
```


















# Session Information

```{r results="verbatim"}
sessionInfo()
```
