---
title: "Clustering of gene profiles"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("config")
```

```{r label="params"}
prefix = "gene_profile_quantile_tss_input"
```


```{r label="blacklist"}
# https://zenbu-wiki.gsc.riken.jp/zenbu/wiki/index.php/Uploading_UCSC_repetitive_elements_track
# downloading UCSC rmsk data as BED
# BED formatted UCSC track content can be obtained from UCSC table broswer.
# The rmsk RepeatMasker (rmsk) track can be exported as BED file by selecting
#
# the assembly "Feb.2009 GRCh37/hg19"
# the group "repeats and variations"
# the track "RepeatMasker"
# and finally the table "rmsk"
# As we desire the complete repetitive elements genome-wide to be loaded into ZENBU, therefore we select
#
# region: "genome"
# ZENBU enable gzip compressed bed files to be loaded directly, so we will further select :
#
# output format: "BED - browser extensibke format"
# output file: we will name the file "UCSC_rmsk.hg19.bed.gz"
# file type returned: "gzip compressed"
options(scipen=999)

if (!exists("rmsk_raw")) {
  rmsk_raw = read.table("~/projects/small_structs/data/rmsk.mm10.bed", skip=1, stringsAsFactors=FALSE)
}
rmsk = rmsk_raw

# blacklist = rmsk
blacklist = rmsk[rmsk[,4] %in% c("GSAT_MM","SYNREP_MM"),]#, "(CTGTG)n"),]#
# blacklist = rbind(blacklist, list("chr19", 61266781, 61266782, "chr19:61266781-61266782", 1404, "+"))
blacklist = rbind(blacklist, list("chr11", 3126500, 3200500, "Sfi1", 300, "-"))
# sat2_feat_saf = rbind(sat2_feat_saf, list("Sfi1", "chr11", 3126500, 3200500, "-"))
print(head(blacklist))
dim(blacklist)
blacklist_filename = "blacklist_file.bed"
write.table(blacklist, file=blacklist_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)    
```






```{r label="tss_feature"}
regions_filenames = c(
  paste0("~/projects/nme2/results/highfatdiet/abs_expr_q4_0.bed"), 
  paste0("~/projects/nme2/results/highfatdiet/abs_expr_q3_0.bed"), 
  paste0("~/projects/nme2/results/highfatdiet/abs_expr_q2_0.bed"), 
  paste0("~/projects/nme2/results/highfatdiet/abs_expr_q1_0.bed"), 
  paste0("~/projects/nme2/results/highfatdiet/abs_expr_q0_0.bed"), 
  NULL
)
```



# Heatmap

```{r, label="computeMatrix tss_keys", eval=TRUE}
matrix_file = paste0("tmp/matrix_", prefix, ".txt.gz")
# score_filenames
score_filenames_scaled = paste0(c(
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/normal_chip_nme2wt_m301_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/normal_chip_nme2wt_m302_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/hghfat_chip_nme2wt_m300_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/hghfat_chip_nme2wt_m303_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/normal_chip_nme2ko_m291_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/normal_chip_nme2ko_m293_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/hghfat_chip_nme2ko_m273_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  "/home/fchuffar/projects/datashare/chip_domenico_tgml/hghfat_chip_nme2ko_m292_end-to-end_trim30_srt_mmq30_PESF_30_4_None_SF",
  NULL
), sprintf("%.2f",df$inv_sf)
, ".bw")

score_filenames = score_filenames_scaled 
brsl = 1500
arsl = 1500
bin_size = 10
prefix = "gene_profile_quantile_tss"
reference_point = "TSS"
matrix_file = deeptoolsr::dt_compute_matrix(regions_filename=regions_filenames, score_filename=score_filenames, out_filename=matrix_file,
  blacklist_filename = blacklist_filename,
  bin_size = bin_size,
  before_region_start_length = brsl,
  after_region_start_length = arsl,
  reference_point = reference_point,
  FORCE_EXEC=FALSE,
  number_of_processors=32,
  sort_regions="descend"
)
source("common.R")
dir.create("tmp", recursive=TRUE, showWarnings=FALSE)
ordered_wig_signal = fast_matrix_file(matrix_file, GRP=TRUE)
hm_out_filename = paste0("tmp/hm_", prefix, ".fasted.png")
hm_out_filename = deeptoolsr::dt_plot_heatmap(matrix_file=paste0(matrix_file, ".fast.txt.gz"), out_filename=hm_out_filename, FORCE_EXEC=FALSE, color_map="YlOrRd", sort_regions="descend")
```

```{r, echo=FALSE, eval=TRUE, out.width="100%", results="markup"}
knitr::include_graphics(hm_out_filename)
```















```{r hira_vs_fact, eval=TRUE}
con = gzfile(matrix_file)
json = readLines(con, n=1)
close(con)
json = substr(json, 2, nchar(json))
metadata = jsonlite::fromJSON(json, simplifyVector = FALSE)
# metadata$sample_boundaries
metadata$group_boundaries

raw_wig_signal = read.table(gzfile(matrix_file), skip=1, stringsAsFactors=FALSE)
rownames(raw_wig_signal) = raw_wig_signal[,4]
head(raw_wig_signal[,1:6])
dim(raw_wig_signal)
mat = as.matrix(raw_wig_signal[,-(1:6)])
head(mat[,1:6])
dim(mat)

delta = (arsl + brsl) / bin_size
foo = lapply(1:(ncol(mat) / delta), function(i) {
  tmp_idx = ((i-1)*delta + 1):(i*delta)
  print(tmp_idx)
  mat[,tmp_idx]
  # apply(mat[,tmp_idx], 1, mean, na.rm=TRUE)
})

mat_normal_wt = (foo[[1]] + foo[[2]]) / 2
mat_hghfat_wt = (foo[[3]] + foo[[4]]) / 2
mat_normal_ko = (foo[[5]] + foo[[6]]) / 2
mat_hghfat_ko = (foo[[7]] + foo[[8]]) / 2


idx_qs = list()
for (q in 0:4) {
  idx_qs[[paste0("q", q)]] = (metadata$group_boundaries[[-q+5]]+1):metadata$group_boundaries[[-q+6]]
}



for (q in 0:4) {
  layout(matrix(1:2, 1), respect=TRUE)
  idx_q = idx_qs[[paste0("q", q)]]

  mat_normal_wt_mean = apply(mat_normal_wt[idx_q,], 2, mean)
  mat_hghfat_wt_mean = apply(mat_hghfat_wt[idx_q,], 2, mean)
  mat_normal_ko_mean = apply(mat_normal_ko[idx_q,], 2, mean)
  mat_hghfat_ko_mean = apply(mat_hghfat_ko[idx_q,], 2, mean)

  plot(0,0,col=0, ylab="average signal (a.u.)", xaxt="n", ylim=c(0,70), xlim=c(0,delta), main=paste0("Q", q, " WT ND (black) HFD (red)"), xlab="")
  axis(1, c(0, delta/2, delta), c("-1.5Kb", "TSS", "+1.5Kb"))  
  lines(mat_normal_wt_mean, type="l", col=1)
  lines(mat_hghfat_wt_mean, type="l", col=2)
  plot(0,0,col=0, ylab="average signal (a.u.)", xaxt="n", ylim=c(0,70), xlim=c(0,delta), main=paste0("Q", q, " KO ND (black) HFD (red)"), xlab="")
  axis(1, c(0, delta/2, delta), c("-1.5Kb", "TSS", "+1.5Kb"))  
  lines(mat_normal_ko_mean, type="l", col=1)
  lines(mat_hghfat_ko_mean, type="l", col=2)
}
```



# Material and Methods



**Sequencing**

Stranded paired-end 2x75bp reads were produced by Illumina NextSeq500. 
Base calling was performed using RTA version 2.*

**Trimming**

Post processing of fastq files, Paired fastq files were processed by 5 prime trimming, keeping 30 bp-length fragments, using `fastx_trimmer`.
with options: -l 30 -Q33, meaning 5 prime trimming keeping 30 bp-length fragments. 

**Alignment**

The trimmed fastq files were aligned on the USCS mm10 genome using `bowtie2`
with options: –end-to-end, –no-mixed, –no-discordant. 

**BigWig production**

The bam siganl were normalized and smoothed using `bamCoverage`
with options: --binSize 4 --minMappingQuality 30 --normalizeUsing RPKM.


**Quartile of expression**

We filtered protein-coding genes from NCBI Reference Sequence Database (RefSeq) and we organise genes by quartile of aboslute DESeq2 normalized expression in WT NFD 4 replicates. Quartile of expression genes were exported as bed files available in supplementary.

**Blacklist**

We exclude from the analysis not only `GSAT_MM` and `SYNREP_MM` tagged loci in repeat masker but also *Sfi1* locus (chr11:3126500-3200500).

**Heatmap**

Heatmap were produced from BigWig files and quartile of expression bed files using `computeMatrix` and `plotHeatmap`
with options: reference-point --referencePoint TSS --binSize 10 --beforeRegionStartLength 500 --afterRegionStartLength 1500 --sortRegions descend.

**TSS Profile**

TSS profiles were generated from `computeMatrix` outputs with a custom R script.






# Session Information

```{r results="verbatim"}
sessionInfo()
date()
```



























