---
title: "TSS profiles protein-coding transcripts"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("config")
source("config.R")
source("common.R")
cores = 1# min(parallel::detectCores(), 32)
```

```{r}
# score_filenames = paste0("/home/chuffarf/projects/datashare/", gse, "/", samples, "_end-to-end_trim30_bowtie2_", species, "_", annotation, "_", version, "_srt.bam")
# bam_file = score_filenames[4]
#   bam_file <- "/home/chuffarf/projects/datashare/atacseq_govin_2023spoatacseq1/veg1_end-to-end_trim30_bowtie2_Saccharomyces_cerevisiae_UCSC_sacCer3_srt.bam"

library(GenomicRanges)
library(VplotR)

# if (!exists("mimportPEBamFiles")) {mimportPEBamFiles = memoise::memoise(importPEBamFiles, cache=cachem::cache_mem(max_size = 10*1024 * 1024^2))}
# if (!exists("mgetFragmentsDistribution")) {mgetFragmentsDistribution = memoise::memoise(getFragmentsDistribution, cache=cachem::cache_mem(max_size = 10*1024 * 1024^2))}
# if (!exists("mplotVmat")) {mplotVmat = memoise::memoise(plotVmat, cache=cachem::cache_mem(max_size = 10*1024 * 1024^2))}


if (!exists("mimportPEBamFiles"))         {mimportPEBamFiles           = importPEBamFiles         }
if (!exists("mgetFragmentsDistribution")) {mgetFragmentsDistribution   = getFragmentsDistribution }
if (!exists("mplotVmat"))                 {mplotVmat                   = plotVmat                 }

for (sample in samples) {
  cat("***********************")
  cat(sample)
  write("***********************", stderr())
  write(sample, stderr())


  bam_file = paste0("/home/chuffarf/projects/datashare/", gse, "/", sample, "_end-to-end_trim30_bowtie2_", species, "_", annotation, "_", version, "_srt.bam")
  # bam_file = "datashare/cutntag_sayou_curie/raw/2021988/analysis_results/N300C04/mapping/N300C04_SC5314_Assembly22_custom_filtered.bam"
  genes_file = paste0("~/projects/datashare/genomes/", species, "/", annotation, "/", version, "/Annotation/Genes/genes.bed")
  genes = read.table(genes_file, header=FALSE)
  head(genes)

  # split and convert per region
  genes = GRanges(
      seqnames = genes$V1,
      ranges = IRanges(start = ifelse(genes$V6=="+", genes$V2, genes$V3),
                      end =   ifelse(genes$V6=="+", genes$V2, genes$V3) + 1,
                      names = genes$V4),
      strand = genes$V6
    )



  fragments = mimportPEBamFiles(
      bam_file, 
      # where=genes,
      # cores=cores,
      shift_ATAC_fragments=TRUE
  )
  # fragments


  df <- mgetFragmentsDistribution(
      fragments, 
      extend_granges = c(-1500, 1500),
      genes, 
      cores=cores
  )
  #> Warning in as.cls(x): NAs introduced by coercion
  #> Warning in as.cls(x): NAs introduced by coercion
  #> Warning in as.cls(x): NAs introduced by coercion
  p <- ggplot(df, aes(x = x, y = y)) + geom_line() + theme_ggplot2() +
       labs(title = sample)
  plot(p)
  #> Warning: 
  p <- mplotVmat(x = fragments,  
    xlims = c(-700, 700),
    # cores=cores, 
    granges = genes
    ) +
       labs(title = sample)
  plot(p)
}
```


# Session Information

```{r results="verbatim"}
sessionInfo()
date()
```



























